runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -1960,6 +1969,79 @@ struct fi_info* getBaseProviderHints(chpl_bool* pTxAttrsForced) {
 }
 
 
+static
+void heedSlingshotSettings(struct fi_info* info) {
+  //
+  // Deal with Slingshot network related settings.
+  //
+  const char* evDevs = getenv("SLINGSHOT_DEVICES");
+  const char* evSvcIds = getenv("SLINGSHOT_SVC_IDS");
+  const char* evTcs = getenv("SLINGSHOT_TCS");
+  const char* evVnis = getenv("SLINGSHOT_VNIS");
+  CHK_TRUE((evDevs == NULL) == (evSvcIds == NULL)
+           && (evSvcIds == NULL) == (evTcs == NULL)
+           && (evTcs == NULL) == (evVnis == NULL)); // sanity
+  if (evDevs == NULL) {
+    return;
+  }
+
+  //
+  // Libfabric allocates and frees struct fi_info structures and their
+  // substructures with the system allocator.  Therefore, that instead
+  // of the Chapel allocator for this auth_key since the pointer to it
+  // will be assigned into that struct.
+  //
+  struct ss_auth_key {
+    uint32_t svc_id;
+    uint16_t vni;
+  };
+  struct ss_auth_key* auth_key;
+  CHK_SYS_MALLOC(auth_key, 1);
+
+  //
+  // Service ID.  If there are more than one, just take the first.
+  //
+  {
+    char ev[strlen(evSvcIds) + 1];  // non-constant, for strtok()
+    strcpy(ev, evSvcIds);
+
+    char* tok;
+    char* lasts;
+    CHK_TRUE((tok = strtok_r(ev, ",", &lasts)) != NULL);
+    CHK_TRUE(sscanf(tok, "%" SCNu32, &auth_key->svc_id) == 1);
+    DBG_PRINTF(DBG_CFG,
+               "Slingshot svc_id %" PRIu32 " (from SVC_IDS=%s)",
+               auth_key->svc_id, evSvcIds);
+  }
+
+  //
+  // VNI.  If there are more than one, just take the first.
```

Comment:
Same.

Reply:
The VNI groups are provided in a comm-separated list parallel to (and 1:1 with) the service IDs. Each comma-separated group may contain up to 4 colon-separated individual VNIs. It's not entirely clear to me why we have multiple VNIs per service ID, but similarly to the service ID case, at may not pertain to Chapel anyway and I felt like the safe way to proceed for now was to just use the first one.